File: scripts/graphzero/zeroref_conformance_gate.py
#!/usr/bin/env python3
"""ZeroRef v1 interoperability conformance gate.

This script is the release/CI gate referenced by bead graphzero-zeroref-v1-shared-cas-1ghi.8.
It blocks broad interoperability claims until the retained three-binary matrix
evidence exists on macOS, Linux, and Windows.

Run from the repo root:
    python3 scripts/zeroref_conformance_gate.py
"""
from __future__ import annotations

import json
import re
import sys
from pathlib import Path
from typing import Any

ROOT = Path(__file__).resolve().parents[2]
EVIDENCE_FILE = ROOT / "docs" / "contracts" / "zeroref-conformance-evidence.json"
CAPABILITY_FIXTURES = ROOT / "docs" / "contracts" / "zeroref-capability-fixtures.json"
README = ROOT / "README.md"
ADR = ROOT / "docs" / "adr" / "002-zeroref-v1.md"

BROAD_PHRASES = [
    "any scheme resolves anywhere",
    "all schemes resolve",
    "resolves across all engines",
    "universal interoperability",
    "any ref works anywhere",
    "every scheme works everywhere",
]
REQUIRED_ENGINES = {"fszero", "graphzero", "tokenzero"}
REQUIRED_OS = {"macos", "linux", "windows"}
HEX = re.compile(r"^[0-9a-f]+$")


def load_json(path: Path) -> Any:
    return json.loads(path.read_text(encoding="utf-8"))


def audit_docs() -> list[str]:
    findings: list[str] = []
    for doc in [README, ADR]:
        if not doc.exists():
            findings.append(f"{doc.name}: missing")
            continue
        text = doc.read_text(encoding="utf-8")
        for line_no, line in enumerate(text.splitlines(), start=1):
            lower = line.lower()
            for phrase in BROAD_PHRASES:
                if phrase in lower:
                    if "<!--" not in line and "evidence" not in lower:
                        findings.append(
                            f"{doc.name}:{line_no}: unmarked broad claim {phrase!r}"
                        )
    return findings


def audit_capability_fixtures() -> list[str]:
    findings: list[str] = []
    if not CAPABILITY_FIXTURES.exists():
        findings.append("capability fixtures missing")
        return findings
    data = load_json(CAPABILITY_FIXTURES)
    peers = data.get("peers", [])
    enabled = [p for p in peers if p.get("name") == "compatible_enabled"]
    disabled = [p for p in peers if p.get("name") == "compatible_but_peer_disabled"]
    if not enabled:
        findings.append("missing compatible_enabled capability fixture")
    if not disabled:
        findings.append("missing compatible_but_peer_disabled capability fixture")
    return findings


def valid_hex(value: Any, length: int) -> bool:
    return isinstance(value, str) and len(value) == length and HEX.fullmatch(value) is not None


def audit_evidence() -> list[str]:
    findings: list[str] = []
    if not EVIDENCE_FILE.exists():
        findings.append(
            f"{EVIDENCE_FILE.name}: missing -- three-binary matrix evidence not yet retained"
        )
        return findings
    data = load_json(EVIDENCE_FILE)
    if data.get("schema") != "zeroref-conformance-evidence/v1":
        findings.append("evidence schema must be 'zeroref-conformance-evidence/v1'")
    for required in ["descriptor_tests", "docs_audit", "timestamp"]:
        if required not in data:
            findings.append(f"evidence file missing field {required}")

    matrix = data.get("matrix")
    if not isinstance(matrix, dict):
        findings.append("evidence matrix must be an object")
        return findings
    status = matrix.get("status")
    if status != "green":
        findings.append(
            f"matrix status is {status!r}; release requires green with macOS/Linux/Windows rows"
        )

    rows = matrix.get("rows")
    if not isinstance(rows, list):
        findings.append("matrix.rows must be an array")
        rows = []
    rows_by_os: dict[str, list[dict[str, Any]]] = {}
    for row in rows:
        if not isinstance(row, dict) or not isinstance(row.get("os"), str):
            findings.append("every matrix row must be an object with an os string")
            continue
        rows_by_os.setdefault(row["os"], []).append(row)
    for os_name in sorted(REQUIRED_OS):
        os_rows = rows_by_os.get(os_name, [])
        if len(os_rows) != 1:
            findings.append(f"matrix requires exactly one {os_name} row, found {len(os_rows)}")
            continue
        cells = os_rows[0].get("cells")
        if not isinstance(cells, list) or not cells:
            findings.append(f"matrix {os_name} row has no cells")
            continue
        non_pass = [cell for cell in cells if not isinstance(cell, dict) or cell.get("status") != "pass"]
        if non_pass:
            findings.append(f"matrix {os_name} row has {len(non_pass)} non-pass cells")
        pairs = {
            (cell.get("writer"), cell.get("reader"))
            for cell in cells
            if isinstance(cell, dict)
        }
        required_pairs = {(writer, reader) for writer in REQUIRED_ENGINES for reader in REQUIRED_ENGINES}
        missing_pairs = sorted(required_pairs - pairs)
        if missing_pairs:
            findings.append(f"matrix {os_name} row misses writer/reader pairs: {missing_pairs}")
        bad_hashes = [
            cell
            for cell in cells
            if isinstance(cell, dict)
            and cell.get("status") == "pass"
            and (
                not valid_hex(cell.get("expected_hash"), 64)
                or cell.get("actual_hash") != cell.get("expected_hash")
            )
        ]
        if bad_hashes:
            findings.append(f"matrix {os_name} row has {len(bad_hashes)} pass cells without byte-exact hashes")

    binaries = matrix.get("sibling_shas")
    if not isinstance(binaries, list):
        findings.append("matrix.sibling_shas must be an array")
        binaries = []
    binary_keys = {
        (entry.get("os"), entry.get("engine"))
        for entry in binaries
        if isinstance(entry, dict)
        and valid_hex(entry.get("commit"), 40)
        and valid_hex(entry.get("sha256"), 64)
    }
    required_binary_keys = {(os_name, engine) for os_name in REQUIRED_OS for engine in REQUIRED_ENGINES}
    missing_binaries = sorted(required_binary_keys - binary_keys)
    if missing_binaries:
        findings.append(f"matrix misses pinned native binaries: {missing_binaries}")
    return findings


def main() -> int:
    findings = []
    findings.extend(audit_docs())
    findings.extend(audit_capability_fixtures())
    findings.extend(audit_evidence())

    if findings:
        print("ZeroRef v1 conformance gate: BLOCKED", file=sys.stderr)
        for f in findings:
            print(f"  - {f}", file=sys.stderr)
        return 1

    print("ZeroRef v1 conformance gate: OK")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tests/graphzero/store/zeroref_conformance_gate.rs
//! ZeroRef v1 interoperability conformance gate (bead 1ghi.8).
//!
//! This test audits GraphZero's public claims against the frozen contract and
//! capability fixtures. It fails when broad interoperability wording appears
//! without an evidence marker, or when the capability descriptor drifts from
//! the runtime constants.
//!
//! The full three-binary matrix gate lives in `scripts/zeroref_conformance_gate.py`,
//! which is wired into release/CI and blocks while macOS/Linux/Windows evidence
//! is absent.

use std::collections::BTreeSet;
use std::fs;
use std::path::PathBuf;

use graphzero_store::store::zeroref_capability::{
    EffectiveState, SharedInteropState, ZeroRefDescriptor, validate_peer_descriptor,
};
use serde::Deserialize;

#[derive(Deserialize)]
struct CapabilityFixture {
    name: String,
    expect: String,
    descriptor: serde_json::Value,
}

#[derive(Deserialize)]
struct CapabilityFixtures {
    peers: Vec<CapabilityFixture>,
}

fn fixture_dir() -> PathBuf {
    PathBuf::from(concat!(env!("CARGO_MANIFEST_DIR"), "/../../docs/contracts"))
}

fn load_capability_fixtures() -> CapabilityFixtures {
    let path = fixture_dir().join("zeroref-capability-fixtures.json");
    let raw = fs::read_to_string(&path).expect("read zeroref-capability-fixtures.json");
    serde_json::from_str(&raw).expect("parse capability fixtures")
}

#[test]
fn descriptor_matches_runtime_constants() {
    let desc = ZeroRefDescriptor::from_env();
    assert_eq!(desc.schema, "zeroref-capability/v1");
    assert_eq!(desc.contract.major, 1);
    assert_eq!(desc.contract.minor, 0);
    assert_eq!(desc.hash.algorithm, "sha256");
    assert_eq!(desc.hash.hex_length, 64);
    assert!(!desc.hash.accept_uppercase);
    assert!(!desc.hash.accept_prefixes);
    assert_eq!(desc.schemes.accepted, vec!["fz", "gz", "tz"]);
    assert_eq!(desc.schemes.emitted, "gz");
    assert_eq!(
        desc.fragments.canonical,
        vec!["#B<start>-<end>", "#L<start>-<end>"]
    );
    assert_eq!(desc.fragments.byte_span, "zero-based-half-open");
    assert_eq!(desc.fragments.line_span, "one-based-inclusive");
    assert!(!desc.fragments.clamps);
    assert_eq!(desc.shared_cas.layout, "blobs/sha256/<hh>/<hash>");
    assert_eq!(desc.shared_cas.layout_version, 1);
    assert!(desc.shared_cas.read);
    assert!(desc.shared_cas.write);
    assert_eq!(desc.shared_cas.max_object_bytes, 256 * 1024 * 1024);
    let classes: BTreeSet<String> = desc.error_classes.iter().cloned().collect();
    let expected: BTreeSet<String> = [
        "malformed",
        "unsupported",
        "range_out_of_bounds",
        "not_utf8",
        "missing",
        "io",
        "digest_mismatch",
        "policy_denied",
        "incompatible_version",
        "legacy_ambiguity",
    ]
    .into_iter()
    .map(|s| s.to_string())
    .collect();
    assert_eq!(classes, expected);
}

#[test]
fn descriptor_passes_enabled_peer_fixture() {
    let fixtures = load_capability_fixtures();
    let ours = ZeroRefDescriptor::current(EffectiveState {
        code_support: true,
        shared_interop: SharedInteropState::Enabled,
        shared_root_configured: true,
        detail: None,
    });
    let enabled = fixtures
        .peers
        .iter()
        .find(|p| p.name == "compatible_enabled")
        .expect("enabled fixture exists");
    assert_eq!(enabled.expect, "compatible");
    let compat =
        validate_peer_descriptor(&enabled.descriptor, &ours).expect("enabled peer is compatible");
    assert!(compat.shared_interop);
}

#[test]
fn descriptor_rejects_disabled_peer_for_shared_interop() {
    let fixtures = load_capability_fixtures();
    let ours = ZeroRefDescriptor::current(EffectiveState {
        code_support: true,
        shared_interop: SharedInteropState::Enabled,
        shared_root_configured: true,
        detail: None,
    });
    let disabled = fixtures
        .peers
        .iter()
        .find(|p| p.name == "compatible_but_peer_disabled")
        .expect("disabled fixture exists");
    let compat = validate_peer_descriptor(&disabled.descriptor, &ours)
        .expect("disabled peer is still compatible at contract level");
    assert!(
        !compat.shared_interop,
        "shared interop must be false when peer is disabled"
    );
}

#[test]
fn docs_do_not_make_unmarked_broad_interop_claims() {
    let root = PathBuf::from(concat!(env!("CARGO_MANIFEST_DIR"), "/../.."));
    let docs = [
        root.join("README.md"),
        root.join("docs/adr/002-zeroref-v1.md"),
        root.join("docs/contracts/zeroref-fixture-cli.md"),
    ];
    let broad_phrases = [
        "any scheme resolves anywhere",
        "all schemes resolve",
        "resolves across all engines",
        "universal interoperability",
        "any ref works anywhere",
        "every scheme works everywhere",
    ];
    for doc in &docs {
        if !doc.exists() {
            continue;
        }
        let text = fs::read_to_string(doc).expect("read doc");
        for (line_no, line) in text.lines().enumerate() {
            let lower = line.to_lowercase();
            for phrase in &broad_phrases {
                if lower.contains(phrase) {
                    assert!(
                        line.contains("<!--") || line.contains("evidence"),
                        "{}:{} contains unmarked broad interoperability claim: {:?}",
                        doc.display(),
                        line_no + 1,
                        phrase
                    );
                }
            }
        }
    }
}
